cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -0,0 +1,192 @@
+// NewCmdApply returns the cobra command for `kubeadm upgrade apply`
+func NewCmdApply(parentFlags *cmdUpgradeFlags) *cobra.Command {
+	flags := &applyFlags{
+		parent:      parentFlags,
+		pullTimeout: 15 * time.Minute,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "apply",
+		Short: "Apply the upgrade",
+		Run: func(cmd *cobra.Command, args []string) {
+			kubeadmutil.CheckErr(ValidateFlags(flags))
+
+			// Default the flags
+			SetFlagDefaults(flags)
+
+			err := RunApply(flags)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+
+	// Specify the valid flags specific for apply
+	cmd.Flags().BoolVarP(&flags.nonInteractiveMode, "yes", "y", flags.nonInteractiveMode, "Perform the upgrade and do not prompt for confirmation (non-interactive mode)")
+	cmd.Flags().BoolVarP(&flags.force, "force", "f", flags.force, "Force upgrading although some requirements might not be met. This also implies non-interactive mode")
+	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be applied. This also implies non-interactive mode")
+	cmd.Flags().StringVar(&flags.newK8sVersionStr, "version", flags.newK8sVersionStr, "Selected version to upgrade to")
+	cmd.Flags().DurationVar(&flags.pullTimeout, "pull-timeout", flags.pullTimeout, "The maximum amount of time to wait for the control plane pods to be downloaded")
+
+	return cmd
+}
+
+// RunApply takes care of the actual upgrade functionality
+func RunApply(flags *applyFlags) error {
+
+	// Start with the basics, verify that the cluster is healthy and get the configuration from the cluster (using the ConfigMap)
+	client, cfg, err := EnforceRequirements(flags.parent.kubeConfigPath, flags.parent.cfgPath, flags.parent.printConfig, flags.nonInteractiveMode)
+	if err != nil {
+		return err
+	}
+
+	// Set the upgraded version on the config object now
+	cfg.KubernetesVersion = flags.newK8sVersionStr
+
+	// Grab the external, versioned configuration and convert it to the internal type for usage here later
+	internalcfg := &kubeadmapi.MasterConfiguration{}
+	api.Scheme.Convert(cfg, internalcfg, nil)
+
+	// Enforce all the version skew policies there are
+	if err := EnforceVersionPolicies(flags, client); err != nil {
+		return err
+	}
+
+	// If interactive mode is set (--yes was not specified), ask ther user whether they really want to upgrade
+	// If --yes is specified; just proceed and do it
+	if !flags.nonInteractiveMode {
+		if err := upgrade.InteractivelyConfirmUpgrade("Are you sure you want to proceed with the upgrade"); err != nil {
+			return err
+		}
+	}
+
+	// Use a prepuller implementation based on creating DaemonSets
+	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
+	prepuller := upgrade.NewDaemonSetPrepuller(client, internalcfg)
+	upgrade.PrepullImagesInParallel(prepuller, flags.pullTimeout)
+
+	// Now; perform the upgrade procedure
+	if err := PerformUpgrade(flags, client, internalcfg); err != nil {
```

Comment:
Based on the implementation, maybe this should be `UpgradeControlPlaneImages()` or something more specific? It was unexpected that `PerformUpgrade()` wasn't actually the full upgrade, and that taking care of things like RBAC/ServiceAccounts/etc. were considered "post-upgrade tasks," which I expected to just be client-side cleanup.